liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/GenerateFeaturesMojo.java

```diff
@@ -0,0 +1,494 @@
+        String classesDirName = null;
+        // First check the Java build output directory (target/classes) for the current project
+        classesDirName = getClassesDirectory(project.getBuild().getOutputDirectory());
+        if (classesDirName != null) {
+            dirs.add(classesDirName);
+        }
+
+        // Use graph to find upstream projects and look for classes directories. Some projects have no Java.
+        ProjectDependencyGraph graph = session.getProjectDependencyGraph();
+        List<MavenProject> upstreamProjects = graph.getUpstreamProjects(project, true);
+        log.debug("For binary scanner gathering Java build output directories for upstream projects, size=" + upstreamProjects.size());
+        for (MavenProject upstreamProject : upstreamProjects) {
+            classesDirName = getClassesDirectory(upstreamProject.getBuild().getOutputDirectory());
+            if (classesDirName != null) {
+                dirs.add(classesDirName);
+            }
+        }
+        for (String s : dirs) {log.debug("Found dir:"+s);};
+        return dirs;
+    }
+
+    // Check one directory and if it exists return its canonical path (or absolute path if error).
+    private String getClassesDirectory(String outputDir) {
+        File classesDir = new File(outputDir);
+        try {
+            if (classesDir.exists()) {
+                return classesDir.getCanonicalPath();
+            }
+        } catch (IOException x) {
+            String classesDirAbsPath = classesDir.getAbsolutePath();
+            log.debug("IOException obtaining canonical path name for a project's classes directory: " + classesDirAbsPath);
+            return classesDirAbsPath;
+        }
+        return null; // directory does not exist.
+    }
+
+    public String getEEVersion(MavenProject project) {
+        List<Dependency> dependencies = project.getDependencies();
+        for (Dependency d : dependencies) {
+            if (!d.getScope().equals("provided")) {
+                continue;
+            }
+            log.debug("getEEVersion, dep="+d.getGroupId()+":"+d.getArtifactId()+":"+d.getVersion());
+            if (d.getGroupId().equals("io.openliberty.features")) {
+                String id = d.getArtifactId();
+                if (id.equals("javaee-7.0")) {
+                    return "ee7";
+                } else if (id.equals("javaee-8.0")) {
+                    return "ee8";
+                } else if (id.equals("javaeeClient-7.0")) {
+                    return "ee7";
+                } else if (id.equals("javaeeClient-8.0")) {
+                    return "ee8";
+                } else if (id.equals("jakartaee-8.0")) {
+                    return "ee8";
+                }
+            } else if (d.getGroupId().equals("jakarta.platform") &&
+                    d.getArtifactId().equals("jakarta.jakartaee-api") &&
+                    d.getVersion().equals("8.0.0")) {
+                return "ee8";
```

Comment:
Jakarta 8 is supported since the features are the same as JavaEE8. Will mention EE9 is not supported as part of https://github.com/OpenLiberty/ci.maven/issues/1337